### PR TITLE
[cli] Disable permalinks to the update details page when using self-managed backends (S3, Azure, GCS)

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #4029
Fixes: #3537

Should the user want to get permalinks when using a self-managed backend, they can pass a flag:

```
$ pulumi up --suppress-permalink false
```

Permalinks for these self-managed backends will be suppressed on `update`, `preview`, `destroy`,
`import` and `refresh` operations.

A view of the functionality:

### self-managed backend

```
$ pulumi whoami -v
User: stack72
Backend URL: file://~


$ PULUMI_CONFIG_PASSPHRASE=password pulumi up
Previewing update (pphrase-2):
     Type                 Name                       Plan
     pulumi:pulumi:Stack  testing-pphrase-2
Resources:
    2 unchanged


$ PULUMI_CONFIG_PASSPHRASE=password pulumi up --suppress-permalink false
Previewing update (pphrase-2):
     Type                 Name                       Plan
     pulumi:pulumi:Stack  testing-pphrase-2
Resources:
    2 unchanged

Permalink: file:///Users/stack72/.pulumi/stacks/pphrase-2.json
```

### Using the SaaS

```
$ pulumi whoami -v
User: stack72
Backend URL: https://app.pulumi.com/stack72


$ pulumi up
Previewing update (dev4)
View Live: https://app.pulumi.com/stack72/testing/dev4/previews/11d3539a-93be-4272-8070-190875b8697f
     Type                          Name                  Plan
 +   pulumi:pulumi:Stack           testing-dev4          create
 +   └─ random:index:RandomString  test                  create
Resources:
    + 2 to create


$ pulumi up --suppress-permalink false
Previewing update (dev4)
View Live: https://app.pulumi.com/stack72/testing/dev4/previews/416a8023-ddcd-4ed1-8d48-5744831b836d
     Type                          Name                  Plan
 +   pulumi:pulumi:Stack           testing-dev4          create
 +   └─ random:index:RandomString  test                  create
Resources:
    + 2 to create


$ pulumi up --suppress-permalink true
Previewing update (dev4)
     Type                          Name                  Plan
 +   pulumi:pulumi:Stack           testing-dev4          create
 +   └─ random:index:RandomString  test                  create
Resources:
    + 2 to create
```

### CODE REVIEW

**Review by infin8x — APPROVED** [review]
Approved from a "does what I expected it to do" perspective. Full code review left to others more qualified :)

File: pkg/cmd/pulumi/destroy.go
Comment (lukehoban) [inline pkg/cmd/pulumi/destroy.go:218]: Two things:
1. We don't use the team "non-saas" anywhere else.  We use "self-managed" in the backends docs, and "filestate" in the implementation.  I'm not sure exactly what's right here - but we should likely base around one of these names instead of introducing "non-saas"
2. We already have `suppress-permalink`.  It feels like this new option should be `suppress-permalink=false`.  We are basically making the defalult for `suppress-permalink` true for the filestate backend with these changes. Not sure exactly how best to represent that with Cobra.

Comment (stack72) [reply pkg/cmd/pulumi/destroy.go:218]: so `suppress-permalink` still works - as we allow suppressing even for hosted backend. I can't make it work with both. The default for this is `false`. Changing this to true gives us chicken and egg based on the hosted backend. I can't figure out a way of being able to use the existing flag without breaking those who use the hosted backend so I was trying to create an explicit flag to make it opt in for self-managed backends

(FWIW, I like selfManagedBackends rather than non-SaaS so thanks for that(

Comment (infin8x) [reply pkg/cmd/pulumi/destroy.go:218]: Ah, the naming is a good point. We're starting to circle around a final set of names for the various backends here: https://docs.google.com/document/d/10cLvyua-xJzVEuZjY0InfLPYeF2RQtG_BLdqUun35D4/edit#heading=h.q32ufnmp2er2

Lemme see if I can draw us to a close there so we can start using those new names.

Comment (infin8x) [reply pkg/cmd/pulumi/destroy.go:218]: Let's stick with `selfManagedBackends` for now - we should have a new name sorted here in the next few weeks, but I don't want to block this PR in the meantime.

**Review by komalali — APPROVED** [review]
LGTM but curious what happens if you just pass in `pulumi up --suppress-permalink`

**Review by lukehoban — APPROVED** [review]

File: pkg/backend/filestate/backend.go
Comment (stack72) [reply pkg/backend/filestate/backend.go:612]: There are 2 places the docs will be updated

The troubleshooting section will have details as to why they get an open bucket url error - it's usually regarding using an Azure and GCp service principal to log into the self managed backends

I will also be updating the state and backends page to talk about the new defaults here of not showing a permalink

Both will be ready early Monday morning PST to make sure they are ready to go

**Comment by fcavaco** [comment]
when trying this in WLS2 , the syntax is slightly different from above if you actually want to add the argument:
--suppress-permalink=false , rather than --suppress-permalink false.

**Pull request merged**
